### DISCUSSION TITLE
Validation using jQuery validate()

### DISCUSSION THREAD

**Discussion by cfcoder** [wheels-dev/wheels#1444] [General]
I'm trying to use jQuery validate plugin in my CFWheels app. I'm not able to get it to work so far. I suspect it has something to do with the startFormTag() in wheels generating a novalidate="novalidate" attribute, but could be wrong. I don't see any way to control that using the startFormTag().  If that is the reason the jQuery validate doesn't work, please tell me how to turn off that attribute. If there is another reason for this jquery validate now working, let me know that.

The reason for this need is I have a form that brings together multiple objects. In the Create action I need to go through a sequence of object.create or object.update calls and each one triggers model validation (wheels generated) for rules I provided in the model's config(). It appears this validation happens first and prevents any jQuery driven validate from working. This presents problems as all the data in the various objects mentioned are related and I need a way to validate all first before trying to step through the list of creates and updates. I'm looking for simple validation like make sure the fields that are required have something in them.

Please advise.

**Comment by bpamiri** [comment]
Bruce,

In addition to the object.create() or object.update() calls, I think the object.save() and object.valid() will be useful in this instance. First of all, the object.save() will automatically determine if an insert or update needs to be run. Second you can call object.valid() which run all the validation on an object without doing any create or update on the object. So in your case call object.valid() on all the secondary objects before doing a object.save() on the main object.

--Peter Amiri
  ***@***.***

**Comment by cfcoder** [comment]
Thank Peter, that looks like it will work quite well for my purposes.
I do have a related question about methods poorly explained in the API Docs.  
What is the difference between addError() and addErrorToBase()?  
Same question for  allErrors() and allErrorsOnBase()?

**Comment by bpamiri** [comment]
The descriptions can probably be modified to make them more clear but addErrors take an additional attribute and allows you to add an error to a specific property, while the addErrorToBase() will add the error on the object and not a specific property in particular.

--Peter Amiri
  ***@***.***

**Comment by cfcoder** [comment]
I don't know what the reference to "Base" is referring to.  AllErrors vs allErrorsOnBase, etc.

**Comment by bpamiri** [comment]
The Base object vs a particular property of the object. Think about a form that has been submitted and validated on the backend. If you want to design the form such that a particular field turns red if it fails validation, then you need to set the error at the property level. But if you simply want to set an error at the form level then you can set the error on the Base.

It just gives you a finer level of control as to what to set the error on and display property specific vs form general error messages.

--Peter Amiri
  ***@***.***

**Reply by cfcoder** [reply]
Thanks Peter. By form level, do you mean object level? Many of my forms include data from multiple models so I have multiple objects involved. As far as I can see, there is no management at a higher "form" level that would collect all object errors for examle, right? I need to call .valid() on each object and display any errors for each. It would be nice if there is a collection of all errors. But even allErrors() is presumed to be called on a per-object basis. So as soon as you involve multiple objects on a form submit, then the management of all validation falls on the user to implement that collection of errors and display of them, correct me if I'm missing something.

**Comment by bpamiri** [comment]
You are correct. I was just saying form since I was talking about form validation not to imply there was a higher level. You may be able to use Object relationships, object validation, and object callbacks to get some more automated functionality but your millage may vary and this will be dependent on your particular circumstances. That is why you may find it easier to just take control of the whole validation process using .valid() on each object.

--Peter Amiri
  ***@***.***

**Comment by cfcoder** [comment]
The docs also don't clearly define which validation methods can be put in the controller and which must be put in the model config() method. I find I can put object.valid() in the controller to pre-check that all is well before a save, but if I put object.validatesPresenceOf() for any property in the object into the controller action, it just returns "undefined".  (https://api.cfwheels.org/v2.5.0/model.validatesPresenceOf)
I can dump the object and the property is clearly defined there but it's empty. But  this check just says it doesn't exist.
